test/types/type-provider.test-d.ts

```diff
@@ -437,3 +436,74 @@ expectAssignable(server.withTypeProvider<JsonSchemaToTsProvider>().get<{Reply: b
     return true
   }
 ))
+
+// -------------------------------------------------------------------
+// FastifyPlugin: Auxiliary
+// -------------------------------------------------------------------
+
+interface AuxiliaryPluginProvider extends FastifyTypeProvider { output: 'plugin-auxiliary' }
+
+// Auxiliary plugins may have varying server types per application. Recommendation would be to explicitly remap instance provider context within plugin if required.
+function plugin<T extends FastifyInstance> (instance: T) {
+  expectAssignable(instance.withTypeProvider<AuxiliaryPluginProvider>().get(
+    '/',
+    {
+      schema: { body: null }
+    },
+    (req) => {
+      expectType<'plugin-auxiliary'>(req.body)
+    }
+  ))
+}
+
+expectAssignable(server.withTypeProvider<AuxiliaryPluginProvider>().register(plugin).get(
+  '/',
+  {
+    schema: { body: null }
+  },
+  (req) => {
+    expectType<'plugin-auxiliary'>(req.body)
+  }
+))
+
+// -------------------------------------------------------------------
+// Handlers: Inline
+// -------------------------------------------------------------------
+
+interface InlineHandlerProvider extends FastifyTypeProvider { output: 'handler-inline' }
+
+// Inline handlers should infer for the request parameters (non-shared)
+expectAssignable(server.withTypeProvider<InlineHandlerProvider>().get(
+  '/',
+  {
+    onRequest: (req, res) => {
+      expectType<'handler-inline'>(req.body)
+    },
+    schema: { body: null }
+  },
+  (req) => {
+    expectType<'handler-inline'>(req.body)
+  }
+))
+
+// -------------------------------------------------------------------
+// Handlers: Auxiliary
+// -------------------------------------------------------------------
+
+interface AuxiliaryHandlerProvider extends FastifyTypeProvider { output: 'handler-auxiliary' }
+
+// Auxiliary handlers are likely shared for multiple routes and thus should infer as unknown due to potential varying parameters
```

Comment:
@RafaelGSS Hi! Yeah, this PR actually makes some distinctions between inline and auxiliary (shared) handlers. I do think documentation conveying some of the reasoning behind the distinction would be good (as it pertains to `unknown`)

Of note, I found it somewhat interesting that the `unknown` sorta falls out naturally of the type system (as TS cannot safely auto infer external functions), but also happens to be the most sensible type for a shared handler. In this sense, TypeScript is providing indirect type safety (causing the user to be explicit about checking values for shared handlers)

This aspect might not be obvious to many users, I think documentation around this would probably bring up the topic of "shared functions may have varying generic parameters and that implementers need to manually check / guard these parameters from `unknown`" or something along those lines.

Reply:
Actually, I think just a section inside https://www.fastify.io/docs/latest/Reference/Type-Providers/ would be enough. Something like:

### Troubleshooting

#### Auxiliary/external handlers won't have types inferred

// Detailed explanation

What do you think?

Reply:
@RafaelGSS Yup :) That sounds good to me. Did you need this documentation included as part of this PR?